src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -50,13 +105,121 @@ rawPageNumBlobs :: RawPage -> Word16
 rawPageNumBlobs (RawPage off ba) = indexByteArray ba (off + 1)
 
 type KeyOffset = Word16
+type ValueOffset = Word16
 
-rawPageKeyOffsets :: RawPage -> Vector KeyOffset
+rawPageKeyOffsets :: RawPage -> P.Vector KeyOffset
 rawPageKeyOffsets page@(RawPage off ba) =
-    Vector (off + fromIntegral (div2 dirOffset)) (fromIntegral dirNumKeys) ba
+    P.Vector (off + fromIntegral (div2 dirOffset)) (fromIntegral dirNumKeys + 1) ba
   where
     Dir {..} = rawPageDirectory page
 
+-- | for non-single key page case
+rawPageValueOffsets :: RawPage -> P.Vector ValueOffset
+rawPageValueOffsets page@(RawPage off ba) =
+    P.Vector (off + fromIntegral (div2 dirOffset) + fromIntegral dirNumKeys) (fromIntegral dirNumKeys + 1) ba
```

Comment:
Right right, offset is `keyOffsets + 2n` (bytes), and length is `2(n+1)`. Hence in the Word16 style, `keyOffsets/2 + n` for offset and `n+1` for length.